### PR TITLE
refactor: Simplify the KubernetesClient DSL

### PR DESCRIPTION
**Pull request by iocanel**
## Description
This pull request simplifies the DSL in the following manner.

- Removes Type parameters when actual parameter never changes:
   - Boolean
   - Watch
- Collapse parameters that refer to the some thing (input, output types).
- Eliminate the need for Doneable.class.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

The PR is still work in progress and the modules currently aligned with the change are:
- kubernetes-client

### CODE REVIEW

**Comment by iocanel** [comment]
@manusa @metacosm: A glimpse of what we discussed.

File: extensions/knative/client/src/main/java/io/fabric8/knative/client/DefaultKnativeClient.java [hunk 1]
```diff
@@ -157,127 +157,127 @@ public NamespacedKnativeClient inNamespace(String namespace) {
...
 
     @Override
-    public MixedOperation<Service, ServiceList, DoneableService, Resource<Service, DoneableService>> services() {
+    public MixedOperation<Service, ServiceList, Resource<Service>> services() {
         return new ServiceOperationsImpl(this.getHttpClient(), this.getConfiguration());
     }
 
     @Override
-    public MixedOperation<Route, RouteList, DoneableRoute, Resource<Route, DoneableRoute>> routes() {
+    public MixedOperation<Route, RouteList, Resource<RouteRoute>> routes() {
```
Comment (metacosm) [inline extensions/knative/client/src/main/java/io/fabric8/knative/client/DefaultKnativeClient.java:165]: That, and subsequent doesn't look right… `RouteRoute`?

File: extensions/service-catalog/client/src/main/resources/resource-operation.vm [hunk 2]
```diff
@@ -53,7 +53,7 @@ import java.util.Map;
 import java.util.TreeMap;
 
 
-public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List, Doneable${model.name}, Resource<${model.name}, Doneable${model.name}>> {
+public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List${model.name}, Resource<${model.name}${model.name}>> {
```
Comment (metacosm) [inline extensions/service-catalog/client/src/main/resources/resource-operation.vm:56]: This should probably be 
```suggestion
public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List${model.name}, Resource<${model.name}>> {
```
instead.

**Comment by metacosm** [comment]
We could probably remove the custom lists as well and replace them by a more generic `KubernetesList<${model.name}>`. What do you think, @manusa @iocanel ?

File: extensions/knative/client/src/main/resources/resource-operation.vm [hunk 3]
```diff
@@ -53,7 +53,7 @@ import java.util.Map;
 import java.util.TreeMap;
 
 
-public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List, Doneable${model.name}, Resource<${model.name}, Doneable${model.name}>> {
+public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List${model.name}, Resource<${model.name}${model.name}>> {
```
Comment (metacosm) [inline extensions/knative/client/src/main/resources/resource-operation.vm:56]: That's also wrong and should be replaced by: 
```suggestion
public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List, Resource<${model.name}${model.name}>> {
```

or, better yet:

```suggestion
public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, KubernetesList<${model.name}>, Resource<${model.name}${model.name}>> {
```

File: extensions/volumesnapshot/client/src/main/resources/resource-operation.vm [hunk 4]
```diff
@@ -53,7 +53,7 @@ import java.util.Map;
 import java.util.TreeMap;
 
 
-public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List, Doneable${model.name}, Resource<${model.name}, Doneable${model.name}>> {
+public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List${model.name}, Resource<${model.name}${model.name}>> {
```
Comment (metacosm) [inline extensions/volumesnapshot/client/src/main/resources/resource-operation.vm:56]:
```suggestion
public class ${model.name}OperationsImpl extends HasMetadataOperation<${model.name}, ${model.name}List, Resource<${model.name}>> {
```

though the same remarks regarding potential simplifications on list and `HasMedataOperation` also apply.

File: openshift-client/src/main/resources/resource-operation.vm [hunk 5]
```diff
@@ -54,7 +54,7 @@ import java.util.Map;
 import java.util.TreeMap;
 
 
-public class ${model.name}OperationsImpl extends OpenShiftOperation<${model.name}, ${model.name}List, Doneable${model.name}, Resource<${model.name}, Doneable${model.name}>> {
+public class ${model.name}OperationsImpl extends OpenShiftOperation<${model.name}, ${model.name}List${model.name}, Resource<${model.name}${model.name}>> {
```
Comment (metacosm) [inline openshift-client/src/main/resources/resource-operation.vm:57]:
```suggestion
public class ${model.name}OperationsImpl extends OpenShiftOperation<${model.name}, ${model.name}List, Resource<${model.name}>> {
```

Same as above.

**Review by metacosm — CHANGES_REQUESTED** [review]
Added some comments on fixes for the duplicated generic types and potential further simplifications.

**Comment by metacosm** [comment]
So it wasn't possible to remove the `*List` classes and simplify the signatures even more?

File: extensions/knative/client/src/main/java/io/fabric8/knative/client/DefaultKnativeClient.java (hunk 1 above)
Comment (iocanel) [reply extensions/knative/client/src/main/java/io/fabric8/knative/client/DefaultKnativeClient.java]: No it doesn't

File: extensions/service-catalog/client/src/main/resources/resource-operation.vm (hunk 2 above)
Comment (iocanel) [reply extensions/service-catalog/client/src/main/resources/resource-operation.vm]: Correct! Fixed along with the other issues.

**Comment by iocanel** [comment]
> So it wasn't possible to remove the `*List` classes and simplify the signatures even more?

I'd rather keep that change out of this pull request. The pull request is already huge and the longer it takes to get it in, the most likely we will end up in `hard to deal with` conflicts.

**Comment by metacosm** [comment]
Is it still possible to pass a `Doneable` if needed?

**Comment by iocanel** [comment]
> Is it still possible to pass a `Doneable` if needed?

No, we no longer generate `Doneable` classes.

**Comment by metacosm** [comment]
> > Is it still possible to pass a `Doneable` if needed?
> 
> No, we no longer generate `Doneable` classes.

OK. That should probably be mentioned in CHANGELOG.md.

What about the tests?

**Comment by iocanel** [comment]
> > > Is it still possible to pass a `Doneable` if needed?
> > 
> > 
> > No, we no longer generate `Doneable` classes.
> 
> OK. That should probably be mentioned in CHANGELOG.md.
> 
> What about the tests?

Working on them.

**Comment by iocanel** [comment]
@rohanKanojia @manusa: I am having trouble reasoning with the integration tests. 

I see compilation errors, that I don't have locally.
Upon closer inspection on the errors, its as if they don't take into consideration the changes introduced in this PR.

For example:

`edit()` changed to `edit(Function<T,T> function)` but the error I get is:

```
/home/runner/work/kubernetes-client/kubernetes-client/kubernetes-itests/src/test/java/io/fabric8/kubernetes/ConfigMapIT.java:[74,101] method edit in interface io.fabric8.kubernetes.client.dsl.Editable<T> cannot be applied to given types;
Error:    required: no arguments
Error:    found: (c)->new C[...]ild()
```

Can you please help?

**Comment by rohanKanojia** [comment]
I recently added caching of maven builds. Maybe it's not working as expected.

**Comment by iocanel** [comment]
> I recently added caching of maven builds. Maybe it's not working as expected.

That makes sense. Adding `-am` when running the integration tests and ensuring we are using the `install` goal could possibly fix that. Is it something you can work on, soon? If not, can you point me to right direction? I have no clue on how the CI stuff work for kubernetes-client these days.

**Comment by rohanKanojia** [comment]
I noticed in your PR that cache wasn't getting saved if it already existed:
```
Post job cleanup.
Cache hit occurred on the primary key Linux-maven-2a2f3d3cb62d425b38f7b65c1ebcc3894e193c62e5c5ac832f94d552117f2269, not saving cache.
```
I've added an additional PR number key which I think should give us a unique cache per pull request in https://github.com/fabric8io/kubernetes-client/pull/2589 . Once it gets merged, maybe the e2e test would start working again.

**Review by manusa — APPROVED** [review]
LGTM, code-wise everything seems to be OK. Thx!

**Comment by iocanel** [comment]
[merge]

**Pull request merged**